vector-im/element-android
# [Voice Broadcast] Use internal playback timer to compute the playback position

## Type of change


## Content

Rely on an internal playback timer to display the current playback position instead of relying on the position given by the android media player.

~Note:
In my opinion, this is not an ideal solution, I reused and improved the existing [CountUpTimer](https://github.com/vector-im/element-android/blob/3acaa16c56d9f843688f1342bb58d1d13f34d373/library/core-utils/src/main/java/im/vector/lib/core/utils/timer/CountUpTimer.kt) which is not based on real elapsed time. Under the hood, it's a `tickerFlow` which notifies at a fixed interval (50 ms in our case) except if we pause it. It can shift a bit for about +/- the `intervalMs` on each pause/resume depending on the timing.
Btw, I noticed that a too-low interval (eg. 10 ms) will break the timer. I also noticed perfs issues after enabling the debugger on android studio, the Flow takes too much time to notify. I don't know if a busy device can have the same kind of issue.~

## Motivation and context

Voice broadcasts sent by Element-Web were causing trouble in the playback position computation.
The chunks sent from the web are encoded in ogg format with aggregated metadata which is not the case for the other VB sent from iOS or Android.

Also, when the recorder is paused on the web, the aggregation is reset, so we cannot rely on the ogg file metadata to calculate the current playback position. The android media player also gives the aggregated position and duration of these files causing glitches in the playback rendering.

I did not find a way to get the relative position in the android MediaPlayer API. Moreover, there are some glitches when we switch to the next file, `MediaPlayer.currentPosition` starts returning the relative position (eg. 0ms/250ms/375ms) and then jumps to the aggregatedPosition (eg. 15_000ms...) but always returns the aggregated duration (eg. 45_000ms for a chunk of 15_000ms, supposing we are listening to the third chunk).

## Screenshots / GIFs





## Tests

- Have several VB with several chunks (you can set the chunk duration to 15s on web in the `/devtools` options) recording from the web and iOS or Android
- Verify that the playback position is correcty updated on each VB
- Verify that the playback position is still correct after moving the cursor
- Stress a bit the playback by switching to another one or playing with the controls (pause/resume)

## Tested devices

- OS version(s):

## Checklist



- [x] Pull request is based on the ~develop~ main branch
